compiler/hash-lower/src/ty/new.rs

```diff
@@ -0,0 +1,263 @@
+
+        // @@Hack: avoid re-creating "commonly" used types in order
+        // to allow for type_id equality to work
+        let id = match ty {
+            IrTy::Char => self.lcx.tys().common_tys.char,
+            IrTy::UInt(UIntTy::U8) => self.lcx.tys().common_tys.u8,
+            IrTy::UInt(UIntTy::U16) => self.lcx.tys().common_tys.u16,
+            IrTy::UInt(UIntTy::U32) => self.lcx.tys().common_tys.u32,
+            IrTy::UInt(UIntTy::U64) => self.lcx.tys().common_tys.u64,
+            IrTy::UInt(UIntTy::U128) => self.lcx.tys().common_tys.u128,
+            IrTy::UInt(UIntTy::USize) => self.lcx.tys().common_tys.usize,
+            IrTy::Int(SIntTy::I8) => self.lcx.tys().common_tys.i8,
+            IrTy::Int(SIntTy::I16) => self.lcx.tys().common_tys.i16,
+            IrTy::Int(SIntTy::I32) => self.lcx.tys().common_tys.i32,
+            IrTy::Int(SIntTy::I64) => self.lcx.tys().common_tys.i64,
+            IrTy::Int(SIntTy::I128) => self.lcx.tys().common_tys.i128,
+            IrTy::Int(SIntTy::ISize) => self.lcx.tys().common_tys.isize,
+            IrTy::Float(FloatTy::F32) => self.lcx.tys().common_tys.f32,
+            IrTy::Float(FloatTy::F64) => self.lcx.tys().common_tys.f64,
+            _ => self.lcx.tys().create(ty),
+        };
+
+        // Add an entry into the cache for this term
+        self.lcx.semantic_cache().borrow_mut().insert(term.into(), id);
+        id
+    }
+
+    /// Get the [IrTy] from the given [TyId].
+    fn ty_from_tir_ty(&self, ty: &Ty) -> IrTy {
+        match ty {
+            Ty::Tuple(TupleTy { data }) => {
+                let mut flags = AdtFlags::empty();
+                flags |= AdtFlags::TUPLE;
+
+                // Convert the single variant into the tuple variant.
+                let variant = self.stores().params().map_fast(*data, |entries| {
+                    let fields = entries.iter().map(|field| self.ty_id_from_tir_ty(field.ty));
+
+                    AdtVariant {
+                        name: "0".into(),
+                        fields: fields
+                            .enumerate()
+                            .map(|(index, ty)| AdtField { name: index.into(), ty })
+                            .collect(),
+                    }
+                });
+
+                let adt = AdtData::new_with_flags("tuple".into(), index_vec![variant], flags);
+                let id = self.lcx.adts().create(adt);
+                IrTy::Adt(id)
+            }
+            Ty::Fn(FnTy { params, return_ty, .. }) => {
+                let params = self.stores().params().map_fast(*params, |params| {
+                    self.lcx.tls().create_from_iter(
+                        params.iter().map(|param| self.ty_id_from_tir_ty(param.ty)),
+                    )
+                });
+
+                let return_ty = self.ty_id_from_tir_ty(*return_ty);
+
+                // @@Todo: functions should have a name if possible, otherwise they default
+                // to an underscore.
```

Comment:
What, like function types?

Reply:
Yes, the function type/definition should store associated attributes, one of them being a name

Reply:
A function type is different from a function definition.